### ISSUE TITLE
Join forces?

### ISSUE THREAD

**Issue by werererer** [inclement/vivarium#98]
Hey I am the core maintainer/Author of japokwm and I have noticed that we share very similar goals with our windowmanagers.

I mean we both used xmonad and didn't keep using it for some reason. Oh and I also got inspired by dwm/sway and somewhat from herbstluft (I initially though you could create layouts in herbstluft lol).

I personally think that my windowmanager has had more development time allocated to it (approx a year). As we share similar goals I think it would be beneficial for us to work on a single project and not to split the development time into two projects. I currently have already put in a year of work and over 1200 commits (100000 lines changed ~10000 remaining lol).
 
### My goals:
- create a windowmanager where you can dynamically create layouts and also dynamically edit them (via lua :D)
- have one workspace stack where workspaces are assigned to monitors similarly to how sway does it
- I also create dwm similar tags instead of workspaces (that took me like 2-3 months because I wanted to improve them :O)
### Where I currently am:
- I already implemented all core features that I want to have
- and xwayland aswell as layershell already work pretty well
- ok xwayland still needs some work :P
- vim like keybindings! (you can chain commands like mod-m+s)
- of course my window manager is also a dynamic tilinc wayland compositor written in c (and lua but mostly in c)
- oh and I have a client: japokmsg that is 99.99% swaymsg but... _don't tell anyone_
- multiple monitors should work rather well (is still a bit unstable)
- and the most important feature: you can create layouts without changing the source code and just by specifying a 3d array (the reset is math) all layouts that are generated in that way are resizable
### Problems
- Is not yet well know. I will work on that though when it is stable enough for normal users ;)
- Sometimes I was tired while programming which resulted in very... interesting naming conventions xD
### Where I need help
- segfaults and other stuff
- unit testing and refactoring
- games don't work that nicely (mouse capture support not that great I messed that one up)
- and a lot of other polishing :D
- fix weirdly named functions
- documentation is yea ok I guess
- My AUR package is not the best of the best but works? and other distros than arch would be great
- fix memory leaks (valgrind for the win)
- I created my icon for the wm in like 2 hours with gimp just to have one it can cretainly be improved.
- since I copied a lot of code together from dwl and sway some stuff is overengineered so yeah more refactoring ... but it works.

yeah and my github page: [japokwm](https://github.com/werererer/japokwm) 

So yeah tell me what you think. I am actually completely ok if you rather want to have an own separate project  :).

**Comment by inclement** [comment]
Hi @werererer, thanks for getting in touch! I knew about japokwm but I hadn't realised how active your development has been and indeed how there are big similarities between our directions. I think back when I started on vivarium that wasn't so obvious - in fact, am I right to think japokwm came from juliawm?

I started writing vivarium as equal parts learning exercise and attempt to solve my own problem (of wanting to try wayland but with a window management behaviour following what I already had with xmonad). Surprisingly it went well enough that it's also core-feature-complete, although I think it's simpler and less ambitious than japokwm. I focused in particular on a simple config format specifically because I felt like it's a better experience for most configuration compared to e.g. xmonad's haskell config, and for the same reason I deliberately don't have any IPC as it stands. I'm not really sure if that was a good idea though and I suspect vivarium is on track to grow some IPC sooner or later! For more dynamic config my own plan has been to provide an easily-wrappable library layer so that it's easy to code layouts or event handling in any language...although that's quite blue sky thinking as it stands.

I think japokwm looks great, but I'd like to continue also working on vivarium - I think there is plenty of space for projects like this with different goals even if they're similar in some ways, and in fact I hope that the window manager ecosystem continues to grow as wlroots makes it easier. Hopefully things like the scene-graph layer make it easier and easier for different compositors to focus on their own unique logic and ideas rather than all the boilerplate which, it seems fair to say, is still somewhat duplicated between them all.

**Comment by werererer** [comment]
Ok thanks for the info :).

Your assumption that japokwm came from juliawm is correct.
I initially wanted to use julia because I wanted to have simple user configuration and I liked it's syntax. Later I realised that it is still too slow so I chose lua instead and had to rename my project because it would have been confusing.

I actually also wanted to have my xmonad config in wayland without the hassle of haskell so I am very open to all features missing from xmonad. And I think that lua can have the potential to be verrry simple as long as the user doesn't need to edit a way to long config file (yeah you are meant awesome).

Btw when you will build in user custom layouts you will inevitably face the problem of resizing custom layouts. I partially fix this problem after a looooot of thinking for all layouts where resizing is dependent on the master window. This doesn't solve the problem for layouts which resize outward from the center of the screen though. I could fix it with an option though that the where the user decides that he wants to resize outward. 

Anyways problems such as these are a reason why I wrote to you because with enough work we may be able to create a tiling manager library from the base code I wrote, where you put in a 3d array, specify the resize direction and amount and it spits out a new 3d array. This library may be then used to write plugins in other programs such as vim/tmux/emacs to enable dwm like tiling windowmanager behavior there. Currently the lack of manpower is hindering this process a lot though...

Also you could actually work on features like enabling coding in external/other languages directly or having very complex event handling (which basically already works in japokwm but can be improved upon) without having to code a lot of the unpleasant code that still awaits you(yeah I am talking of you ipc) and we both would get the windowmanager that we like to have faster.

Also there currently is a lack of a standardized ipc protocol to communicate workspaces to bars like ewmh(X11) and we then could focus on such things more.

In other words working together may enable us to create an even better window manager than doing alone would allow us.

Anyways if that is still not convincing enough for you then so be it :D. I mean more window managers = more innovation in features :).
